qemu/tests/watchdog.py

```diff
@@ -89,12 +89,42 @@ def _action_check(test, session, watchdog_action):
         """
         # when watchdog action is pause, shutdown, reset, poweroff
         # the vm session will lost responsive
+
+        def check_guest_reboot(pattern):
+            start_time = time.time()
+            while (time.time() - start_time) < vm.REBOOT_TIMEOUT:
+                if pattern in vm.serial_console.get_output():
+                    return True
+            return False
+
         response_timeout = int(params.get("response_timeout", '240'))
+        vm_arch_name = params["vm_arch_name"]
         error_context.context("Check whether or not watchdog action '%s' took"
                               " effect" % watchdog_action, logging.info)
+        if watchdog_action == "inject-nmi":
+            if (vm_arch_name in ("x86_64", "i686")):
+                if not utils_misc.wait_for(lambda: "NMI received" in session.cmd_output("dmesg"),
+                                           response_timeout, 0, 1):
+                    test.fail("Guest didn't receive dmesg with 'NMI received',"
+                              "after action '%s'." % watchdog_action)
+                msg = session.cmd_output("dmesg").splitlines()[-8:]
+                logging.info("Guest received dmesg info: %s" % msg)
+            elif (vm_arch_name in ("ppc64", "ppc64le")):
+                vm.monitor.clear_event("WATCHDOG")
```

Comment:
If the action takes effect too fast, you'll receive the event before you clean it, if so, it'll never pass the next check.